### ISSUE TITLE
distutils/ccompiler.py:68 UnicodeEncodeError on Python 3.2 (migrated from Trac #1601)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3152]
Original ticket http://projects.scipy.org/numpy/ticket/1601
Reported 2010-09-02 by trac user srid, assigned to atmention:cournape.

Python 3.2 py3k branch | Linux 64-bit (FC4) | GCC 4.0.x | numpy-1.5.0 | ATLAS

```
Running from numpy source directory.Traceback (most recent call last):
  File "setup.py", line 211, in <module>
    setup_package()
  File "setup.py", line 204, in setup_package
    configuration=configuration )
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/core.py", line 186, in setup
    return old_setup(**new_attr)
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/core.py", line 149, in setup
    dist.run_commands()
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/dist.py", line 919, in run_commands
    self.run_command(cmd)
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/dist.py", line 938, in run_command
    cmd_obj.run()
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/build.py", line 37, in run
    old_build.run(self)
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/command/build.py", line 128, in run
    self.run_command(cmd_name)
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/cmd.py", line 315, in run_command
    self.distribution.run_command(command)
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/dist.py", line 938, in run_command
    cmd_obj.run()
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/build_src.py", line 152, in run
    self.build_sources()
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/build_src.py", line 169, in build_sources
    self.build_extension_sources(ext)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/build_src.py", line 328, in build_extension_sources
    sources = self.generate_sources(sources, ext)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/build_src.py", line 385, in generate_sources
    source = func(extension, build_dir)
  File "numpy/core/setup.py", line 399, in generate_config_h
    moredefs, ignored = cocache.check_types(config_cmd, ext, build_dir)
  File "numpy/core/setup.py", line 41, in check_types
    out = check_types(*a, **kw)
  File "numpy/core/setup.py", line 294, in check_types
    res = config_cmd.check_type_size(complex_def, expected=2*expected[type])
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/config.py", line 236, in check_type_size
    headers, include_dirs, 'c')
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/config.py", line 99, in _compile
    (body, headers, include_dirs, lang))
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/command/config.py", line 89, in _wrap_method
    ret = mth(*((self,)+args))
  File "/opt/ActivePython-3.2/lib/python3.2/distutils/command/config.py", line 135, in _compile
    self.compiler.compile([src], include_dirs=include_dirs)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/ccompiler.py", line 33, in <lambda>
    m = lambda self, *args, **kw: func(self, *args, **kw)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/ccompiler.py", line 206, in CCompiler_compile
    self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/ccompiler.py", line 33, in <lambda>
    m = lambda self, *args, **kw: func(self, *args, **kw)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/unixccompiler.py", line 23, in UnixCCompiler__compile
    extra_postargs, display = display)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/ccompiler.py", line 33, in <lambda>
    m = lambda self, *args, **kw: func(self, *args, **kw)
  File "/tmp/tmpJXOWDV-pypm-numpy-1.5.0/numpy-1.5.0/build/py3k/numpy/distutils/ccompiler.py", line 68, in CCompiler_spawn
    print(o)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 50-52: ordinal not in range(128)
```

(BTW, there is no 1.5.0 version in 'Version' list)

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-01

{{{exec_command}}} returns output in latin1 encoding, so have to use the {{{asstr}}} function from numpy.compat before using it. I think this fixes it: https://github.com/rgommers/numpy/tree/ticket-1601-py32 Does that work for you?

**Comment by thouis** [comment]
Comment in Trac by trac user srid, 2011-03-02

I cannot seem to reproduce this with numpy-1.5.1
http://code.activestate.com/pypm/numpy/

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

When you encountered the error, did you use 1.5.0 final? If not, it was probably addressed by commit:4e2e78f9 It is also possible that this was a python 3.2 bug.

The o/out argument to print should already be a string, so my patch actually does nothing.

It seems to work in 1.5.1 and trunk, so ticket can be closed?

**Comment by thouis** [comment]
Comment in Trac by trac user srid, 2011-03-03

Hmm, actually I was wrong. I seem to have already worked around this issue with the following (hackish) patch:

# Index: numpy-1.5.0/numpy/distutils/ccompiler.py

--- numpy-1.5.0.orig/numpy/distutils/ccompiler.py   2010-08-21 22:08:35.000000000 -0700
+++ numpy-1.5.0/numpy/distutils/ccompiler.py    2010-09-10 17:32:45.000000000 -0700
atmention:atmention: -65,7 +65,7 atmention:atmention:
     if s:
         if is_sequence(cmd):
             cmd = ' '.join(list(cmd))
-        print(o)
-        print(o.encode('utf8') if sys.version_info[0] >= 3 else o)
       if re.search('Too many open files', o):
           msg = '\nTry rerunning setup command until build succeeds.'
       else:

I suppose this ticket can be closed anyway.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

Wait, without your patch this is still failing for you then? Then we can't close it.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2012-03-26

This was fixed a while ago in commit:dba98cc143, the code now looks like
            try:
                print(o)
            except UnicodeError:
